torch/jit/__init__.py

```diff
@@ -702,32 +682,100 @@ def init_then_register(self, *args, **kwargs):
         return super(ScriptMeta, cls).__init__(name, bases, attrs)
 
 
-class ScriptModule(with_metaclass(ScriptMeta, torch._C.ScriptModule)):
-
-    def __setattr__(self, name, value):
-        if isinstance(value, Parameter):
-            self._register_or_set_parameter(name, value)
-        elif isinstance(value, ScriptModule):
-            self._register_module(name, value)
-            # note: script modules are subclassed in python and the
-            # C++ script::Module class will not hold references to them
-            # to ensure that you always get the same python value here
-            # we store it as a native attribute _in addition to_
-            # registering it with the C++ script::Module
-            object.__setattr__(self, name, value)
-        else:
-            object.__setattr__(self, name, value)
+class ScriptModule(with_metaclass(ScriptMeta, Module, torch._C.ScriptModule)):
+    def __init__(self, optimize=True):
+        Module.__init__(self)
+        torch._C.ScriptModule.__init__(self, optimize)
```

Comment:
We should make sure that forgetting to call `super().__init__()` in a user-defined subclass can't cause a segfault. I've seen it happen multiple times with `pybind`, because it doesn't construct the C++ object before `__init__` (it should probably do that in `__new__`, but oh well)

Reply:
I moved the call to ._C.ScriptModule into the init that is generated by the metaclass, so it gets called even if super is not called.